core/src/main/java/org/elasticsearch/transport/netty/MessageChannelHandler.java

```diff
@@ -342,4 +313,66 @@ public void onFailure(Throwable e) {
             }
         }
     }
+
+    /**
+     * Internal helper class to store characteristic offsets of a buffer during processing
+     */
+    protected static final class Marker {
+        private final ChannelBuffer buffer;
+        private final int size;
+        private final int markedReaderIndex;
+        private final int expectedReaderIndex;
+
+        public Marker(ChannelBuffer buffer) {
+            this.buffer = buffer;
+            this.size = buffer.getInt(buffer.readerIndex() - 4);
```

Comment:
Oh, sorry I totally missed your previous comment... . I look at it, add more docs and push another commit.